vizro-ai/examples/dashboard_ui/actions.py

```diff
@@ -0,0 +1,131 @@
+"""Custom actions used within a dashboard."""
+
+import base64
+import io
+import logging
+
+import black
+import pandas as pd
+from _utils import check_file_extension
+from dash.exceptions import PreventUpdate
+from langchain_openai import ChatOpenAI
+from plotly import graph_objects as go
+from vizro.models.types import capture
+from vizro_ai import VizroAI
+
+SUPPORTED_VENDORS = {"OpenAI": ChatOpenAI}
+
+
+def get_vizro_ai_plot(user_prompt, df, model, api_key, api_base, vendor_input):  # noqa: PLR0913
+    """VizroAi plot configuration."""
+    vendor = SUPPORTED_VENDORS[vendor_input]
+    llm = vendor(model_name=model, openai_api_key=api_key, openai_api_base=api_base)
+    vizro_ai = VizroAI(model=llm)
+    ai_outputs = vizro_ai.plot(df, user_prompt, explain=False, return_elements=True)
+
+    return ai_outputs
+
+
+@capture("action")
+def run_vizro_ai(user_prompt, n_clicks, data, model, api_key, api_base, vendor_input):  # noqa: PLR0913
+    """Gets the AI response and adds it to the text window."""
+
+    def create_response(ai_response, figure, user_prompt, filename):
+        plotly_fig = figure.to_json()
+        return (
+            ai_response,
+            figure,
+            {"ai_response": ai_response, "figure": plotly_fig, "prompt": user_prompt, "filename": filename},
+        )
+
+    if not n_clicks:
+        raise PreventUpdate
+
+    if not data:
+        ai_response = "Please upload data to proceed!"
+        figure = go.Figure()
+        return create_response(ai_response, figure, user_prompt, None)
+
+    if not api_key:
+        ai_response = "API key not found. Make sure you enter your API key!"
+        figure = go.Figure()
+        return create_response(ai_response, figure, user_prompt, data["filename"])
+
+    if api_key.startswith('"'):
+        ai_response = "Make sure you enter your API key without quotes!"
+        figure = go.Figure()
+        return create_response(ai_response, figure, user_prompt, data["filename"])
+
+    if api_base is not None and api_base.startswith('"'):
+        ai_response = "Make sure you enter your API base without quotes!"
+        figure = go.Figure()
+        return create_response(ai_response, figure, user_prompt, data["filename"])
+
+    try:
+        df = pd.DataFrame(data["data"])
+        ai_outputs = get_vizro_ai_plot(
+            user_prompt=user_prompt,
+            df=df,
+            model=model,
+            api_key=api_key,
+            api_base=api_base,
+            vendor_input=vendor_input,
+        )
+        ai_code = ai_outputs.code
+        figure = ai_outputs.figure
+        formatted_code = black.format_str(ai_code, mode=black.Mode(line_length=100))
+
+        ai_response = "\n".join(["```python", formatted_code, "```"])
+        return create_response(ai_response, figure, user_prompt, data["filename"])
+
+    except Exception as exc:
+        logging.exception(exc)
+        ai_response = f"Sorry, I can't do that. Following Error occurred: {exc}"
```

Comment:
Possible future change: might be worth putting the `exc` in a code block so it renders better but don't do this without testing carefully because it might make rendering worse...